In [3]:
import os
import zipfile

ZIP_PATH = "/Users/komal/Documents/SHL/shl-hiring-assessment-2026.zip"
EXTRACT_DIR = "/Users/komal/Documents/SHL/shl_dataset"

os.makedirs(EXTRACT_DIR, exist_ok=True)

with zipfile.ZipFile(ZIP_PATH, "r") as zip_ref:
    zip_ref.extractall(EXTRACT_DIR)

print("Dataset extracted successfully!")
print("Location:", EXTRACT_DIR)

Dataset extracted successfully!
Location: /Users/komal/Documents/SHL/shl_dataset


In [1]:
import shutil

print("FFmpeg path:", shutil.which("ffmpeg"))

FFmpeg path: /opt/homebrew/bin/ffmpeg


In [4]:
for root, dirs, files in os.walk(EXTRACT_DIR):
    level = root.replace(EXTRACT_DIR, "").count(os.sep)
    indent = "    " * level
    print(f"{indent}{os.path.basename(root)}/")
    
    if level < 2:
        for file in files[:10]:
            print(f"{indent}    {file}")

shl_dataset/
    .DS_Store
    Dataset_Final/
        .DS_Store
        test.csv
        train.csv
        sample_submission.csv
        test/
        train/


In [5]:
import glob

print("Train CSV:")
print(glob.glob(EXTRACT_DIR + "/**/train.csv", recursive=True))

print("\nTest CSV:")
print(glob.glob(EXTRACT_DIR + "/**/test.csv", recursive=True))

print("\nSample submission:")
print(glob.glob(EXTRACT_DIR + "/**/sample_submission.csv", recursive=True))

Train CSV:
['/Users/komal/Documents/SHL/shl_dataset/Dataset_Final/train.csv']

Test CSV:
['/Users/komal/Documents/SHL/shl_dataset/Dataset_Final/test.csv']

Sample submission:
['/Users/komal/Documents/SHL/shl_dataset/Dataset_Final/sample_submission.csv']


In [6]:
import os
import pandas as pd

BASE_DIR = "/Users/komal/Documents/SHL/shl_dataset/Dataset_Final"

TRAIN_CSV = os.path.join(BASE_DIR, "train.csv")
TEST_CSV = os.path.join(BASE_DIR, "test.csv")
SAMPLE_SUBMISSION = os.path.join(BASE_DIR, "sample_submission.csv")

TRAIN_AUDIO_DIR = os.path.join(BASE_DIR, "train")
TEST_AUDIO_DIR = os.path.join(BASE_DIR, "test")

print("Train CSV exists:", os.path.exists(TRAIN_CSV))
print("Test CSV exists:", os.path.exists(TEST_CSV))
print("Sample submission exists:", os.path.exists(SAMPLE_SUBMISSION))
print("Train audio exists:", os.path.exists(TRAIN_AUDIO_DIR))
print("Test audio exists:", os.path.exists(TEST_AUDIO_DIR))

Train CSV exists: True
Test CSV exists: True
Sample submission exists: True
Train audio exists: True
Test audio exists: True


In [7]:
train_df = pd.read_csv(TRAIN_CSV)
test_df = pd.read_csv(TEST_CSV)
sample_df = pd.read_csv(SAMPLE_SUBMISSION)

print("Train shape:", train_df.shape)
print("Test shape :", test_df.shape)
print("Sample shape:", sample_df.shape)

print("\nTrain columns:")
print(train_df.columns.tolist())

print("\nTest columns:")
print(test_df.columns.tolist())

print("\nSample submission columns:")
print(sample_df.columns.tolist())

display(train_df.head())
display(test_df.head())
display(sample_df.head())

Train shape: (769, 2)
Test shape : (216, 2)
Sample shape: (204, 2)

Train columns:
['filename', 'label']

Test columns:
['filename', 'label']

Sample submission columns:
['filename', 'label']


,filename,label
0,audio_192.wav,3.0
1,audio_436.wav,3.0
2,audio_447.wav,3.5
3,audio_126.wav,2.0
4,audio_61.wav,2.0


,filename,label
0,audio_128.wav,-1
1,audio_156.wav,-1
2,audio_19.wav,-1
3,audio_108.wav,-1
4,audio_206.wav,-1


,filename,label
0,audio_804.wav,-1
1,audio_1028.wav,-1
2,audio_865.wav,-1
3,audio_774.wav,-1
4,audio_1138.wav,-1


In [8]:
train_audio_files = os.listdir(TRAIN_AUDIO_DIR)
test_audio_files = os.listdir(TEST_AUDIO_DIR)

print("Number of train audio files:", len(train_audio_files))
print("Number of test audio files :", len(test_audio_files))

print("\nExample train files:")
print(train_audio_files[:5])

print("\nExample test files:")
print(test_audio_files[:5])

Number of train audio files: 769
Number of test audio files : 216

Example train files:
['audio_735.wav', 'audio_721.wav', 'audio_709.wav', 'audio_5037.wav', 'audio_292.wav']

Example test files:
['audio_127.wav', 'audio_70.wav', 'audio_133.wav', 'audio_64.wav', 'audio_58.wav']


In [9]:
!pip install -U openai-whisper

In [10]:
import whisper

print("Whisper imported successfully!")

Whisper imported successfully!


In [11]:
whisper_model = whisper.load_model("base")

print("Whisper base model loaded successfully!")

Whisper base model loaded successfully!


In [12]:
print(train_df.columns.tolist())
print(train_df.head())

['filename', 'label']
        filename  label
0  audio_192.wav    3.0
1  audio_436.wav    3.0
2  audio_447.wav    3.5
3  audio_126.wav    2.0
4   audio_61.wav    2.0


In [13]:
filename = train_df["filename"].iloc[0]
audio_path = os.path.join(TRAIN_AUDIO_DIR, filename)

print("Filename:", filename)
print("Audio exists:", os.path.exists(audio_path))

result = whisper_model.transcribe(
    audio_path,
    fp16=False
)

print("\nTranscript:")
print(result["text"])

Filename: audio_192.wav
Audio exists: True

Transcript:
 Well, this petite land is very, very beautiful and fun. There are fleets like the navy around the swing. The swing usually enjoys this swing. It has to be like, it's more like a game of play. Because once it's in it and the other thing, we'll have to finish and the miracle round, which is fine for me. And most of them play for some parts. Why others use these swings? Some of the people like to do sound classes. Most times, here the sound of stringent children, lasting kids.


In [14]:
result = whisper_model.transcribe(
    audio_path,
    fp16=False
)

print("Transcript:")
print(result["text"])

Transcript:
 Well, this petite land is very, very beautiful and fun. There are fleets like the navy around the swing. The swing usually enjoys this swing. It has to be like, it's more like a game of play. Because once it's in it and the other thing, we'll have to finish and the miracle round, which is fine for me. And most of them play for some parts. Why others use these swings? Some of the people like to do sound classes. Most times, here the sound of stringent children, lasting kids.


In [15]:
import os
import pandas as pd
from tqdm.notebook import tqdm

sample_transcripts = []

for _, row in tqdm(train_df.head(10).iterrows(), total=10, desc="Transcribing samples"):
    filename = row["filename"]
    audio_path = os.path.join(TRAIN_AUDIO_DIR, filename)

    result = whisper_model.transcribe(
        audio_path,
        fp16=False
    )

    sample_transcripts.append({
        "filename": filename,
        "label": row["label"],
        "transcript": result["text"].strip()
    })

sample_transcripts_df = pd.DataFrame(sample_transcripts)

display(sample_transcripts_df)

Transcribing samples:   0%|          | 0/10 [00:00<?, ?it/s]

,filename,label,transcript
0,audio_192.wav,3.0,"Well, this petite land is very, very beautiful..."
1,audio_436.wav,3.0,"To see no of a school playground, our playgrou..."
2,audio_447.wav,3.5,I favorite place to visit is the beach. I love...
3,audio_126.wav,2.0,"Once upon a time, we, with five friends, went ..."
4,audio_61.wav,2.0,"Well, when I was a child, I remember I went to..."
5,audio_639.wav,2.5,The best thing of my life was probably when I ...
6,audio_703.wav,2.0,I don't know. Do you want to talk to me? I nev...
7,audio_509.wav,3.0,is smoothing effects. Besides this I also have...
8,audio_672.wav,4.0,The school playground is very big and is very ...
9,audio_344.wav,4.0,My normal goal in life is to become the best o...


In [16]:
for _, row in sample_transcripts_df.iterrows():
    print("=" * 80)
    print("Filename :", row["filename"])
    print("Grammar score:", row["label"])
    print("Transcript:")
    print(row["transcript"])
    print()

Filename : audio_192.wav
Grammar score: 3.0
Transcript:
Well, this petite land is very, very beautiful and fun. There are fleets like the navy around the swing. The swing usually enjoys this swing. It has to be like, it's more like a game of play. Because once it's in it and the other thing, we'll have to finish and the miracle round, which is fine for me. And most of them play for some parts. Why others use these swings? Some of the people like to do sound classes. Most times, here the sound of stringent children, lasting kids.

Filename : audio_436.wav
Grammar score: 3.0
Transcript:
To see no of a school playground, our playground looks like a very big playground having a more more greenery trees in which we students like to garden a more trees to have a fresh year and all having a fresh oxygen. This in the playground, these trees gives a very greenery view which will very present to our minds and will give relaxation. We students like to like the activities of games like cocoa, voli

In [17]:
import os
import pandas as pd
from tqdm.notebook import tqdm

TRANSCRIPT_DIR = "/Users/komal/Documents/SHL/transcripts"
os.makedirs(TRANSCRIPT_DIR, exist_ok=True)

TRAIN_TRANSCRIPT_CSV = os.path.join(
    TRANSCRIPT_DIR, "train_transcripts.csv"
)

TEST_TRANSCRIPT_CSV = os.path.join(
    TRANSCRIPT_DIR, "test_transcripts.csv"
)

print("Transcript directory:", TRANSCRIPT_DIR)

Transcript directory: /Users/komal/Documents/SHL/transcripts


In [18]:
import os
import pandas as pd
from tqdm.notebook import tqdm

if os.path.exists(TRAIN_TRANSCRIPT_CSV):
    train_transcripts_df = pd.read_csv(TRAIN_TRANSCRIPT_CSV)
    processed_files = set(train_transcripts_df["filename"])
else:
    train_transcripts_df = pd.DataFrame(
        columns=["filename", "label", "transcript"]
    )
    processed_files = set()

for _, row in tqdm(
    train_df.iterrows(),
    total=len(train_df),
    desc="Transcribing training audio"
):
    filename = row["filename"]

    if filename in processed_files:
        continue

    audio_path = os.path.join(TRAIN_AUDIO_DIR, filename)

    try:
        result = whisper_model.transcribe(
            audio_path,
            fp16=False
        )

        new_row = pd.DataFrame([{
            "filename": filename,
            "label": row["label"],
            "transcript": result["text"].strip()
        }])

        train_transcripts_df = pd.concat(
            [train_transcripts_df, new_row],
            ignore_index=True
        )

        train_transcripts_df.to_csv(
            TRAIN_TRANSCRIPT_CSV,
            index=False
        )

    except Exception as e:
        print(f"\nError processing {filename}: {e}")

print("\nTraining transcription completed.")
print("Number of transcripts:", len(train_transcripts_df))

Transcribing training audio:   0%|          | 0/769 [00:00<?, ?it/s]


Training transcription completed.
Number of transcripts: 769


In [19]:
if os.path.exists(TEST_TRANSCRIPT_CSV):
    test_transcripts_df = pd.read_csv(TEST_TRANSCRIPT_CSV)
    processed_files = set(test_transcripts_df["filename"])
else:
    test_transcripts_df = pd.DataFrame(
        columns=["filename", "transcript"]
    )
    processed_files = set()

for _, row in tqdm(
    test_df.iterrows(),
    total=len(test_df),
    desc="Transcribing test audio"
):
    filename = row["filename"]

    if filename in processed_files:
        continue

    audio_path = os.path.join(TEST_AUDIO_DIR, filename)

    try:
        result = whisper_model.transcribe(
            audio_path,
            fp16=False
        )

        new_row = pd.DataFrame([{
            "filename": filename,
            "transcript": result["text"].strip()
        }])

        test_transcripts_df = pd.concat(
            [test_transcripts_df, new_row],
            ignore_index=True
        )

        test_transcripts_df.to_csv(
            TEST_TRANSCRIPT_CSV,
            index=False
        )

    except Exception as e:
        print(f"\nError processing {filename}: {e}")

print("\nTest transcription completed.")
print("Number of transcripts:", len(test_transcripts_df))

Transcribing test audio:   0%|          | 0/216 [00:00<?, ?it/s]


Test transcription completed.
Number of transcripts: 216


In [20]:
print("Train transcripts:", len(train_transcripts_df))
print("Test transcripts :", len(test_transcripts_df))

display(train_transcripts_df.head())
display(test_transcripts_df.head())

Train transcripts: 769
Test transcripts : 216


,filename,label,transcript
0,audio_192.wav,3.0,"Well, this petite land is very, very beautiful..."
1,audio_436.wav,3.0,"To see no of a school playground, our playgrou..."
2,audio_447.wav,3.5,I favorite place to visit is the beach. I love...
3,audio_126.wav,2.0,"Once upon a time, we, with five friends, went ..."
4,audio_61.wav,2.0,"Well, when I was a child, I remember I went to..."


,filename,transcript
0,audio_128.wav,The moment in my life that I cherish the most ...
1,audio_156.wav,One of the role models is someone who embadiys...
2,audio_19.wav,"I guess my favorite sport would be Taekwondo, ..."
3,audio_108.wav,Hospital having more of doctors and there are ...
4,audio_206.wav,"I remember when I went to Goa for a trip, then..."


In [22]:
!pip install -q spacy

In [23]:
!{__import__('sys').executable} -m spacy download en_core_web_sm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 19.9 MB/s  0:00:001.1 MB/s eta 0:00:01
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [24]:
import spacy

nlp = spacy.load("en_core_web_sm")

print("spaCy loaded successfully!")

spaCy loaded successfully!


In [26]:
import spacy

nlp = spacy.load("en_core_web_sm")

print("spaCy loaded successfully.")

spaCy loaded successfully.


In [29]:
import re
import numpy as np
import pandas as pd

FILLER_WORDS = {
    "um", "uh", "erm", "hmm", "like",
    "you know", "actually", "basically"
}

SELF_CORRECTION_PHRASES = {
    "i mean",
    "or rather",
    "sorry",
    "what i mean",
    "let me rephrase",
    "i mean to say"
}


def extract_advanced_linguistic_features(text):
    text = str(text).strip()
    doc = nlp(text)

    features = {}

    words = [
        token.text.lower()
        for token in doc
        if token.is_alpha
    ]

    word_count = len(words)
    sentences = list(doc.sents)
    sentence_count = len(sentences)

    # --------------------------------------------------
    # Basic statistics
    # --------------------------------------------------

    features["word_count"] = word_count
    features["unique_word_count"] = len(set(words))

    features["type_token_ratio"] = (
        len(set(words)) / word_count
        if word_count else 0
    )

    features["sentence_count"] = sentence_count

    sentence_lengths = []

    for sent in sentences:
        sent_words = [
            token for token in sent
            if token.is_alpha
        ]
        sentence_lengths.append(len(sent_words))

    features["avg_sentence_length"] = (
        np.mean(sentence_lengths)
        if sentence_lengths else 0
    )

    features["sentence_length_std"] = (
        np.std(sentence_lengths)
        if sentence_lengths else 0
    )

    features["max_sentence_length"] = (
        max(sentence_lengths)
        if sentence_lengths else 0
    )

    features["min_sentence_length"] = (
        min(sentence_lengths)
        if sentence_lengths else 0
    )

    # --------------------------------------------------
    # POS distribution
    # --------------------------------------------------

    pos_counts = {}

    for token in doc:
        if token.is_alpha:
            pos_counts[token.pos_] = (
                pos_counts.get(token.pos_, 0) + 1
            )

    for pos in [
        "NOUN", "VERB", "AUX", "ADJ",
        "ADV", "PRON", "DET", "ADP",
        "CCONJ", "SCONJ", "NUM"
    ]:

        count = pos_counts.get(pos, 0)

        features[f"{pos.lower()}_count"] = count

        features[f"{pos.lower()}_ratio"] = (
            count / word_count
            if word_count else 0
        )

    # --------------------------------------------------
    # Verb complexity
    # --------------------------------------------------

    verbs = [
        token for token in doc
        if token.pos_ == "VERB"
    ]

    finite_verbs = [
        token for token in doc
        if token.pos_ in {"VERB", "AUX"}
        and token.morph.get("VerbForm")
    ]

    features["verb_count"] = len(verbs)

    features["finite_verb_count"] = len(finite_verbs)

    features["verbs_per_sentence"] = (
        len(verbs) / sentence_count
        if sentence_count else 0
    )

    # --------------------------------------------------
    # Clause / sentence complexity
    # --------------------------------------------------

    subordinate_count = sum(
        1 for token in doc
        if token.dep_ in {
            "advcl",
            "ccomp",
            "xcomp",
            "acl",
            "relcl"
        }
    )

    features["subordinate_clause_count"] = subordinate_count

    features["subordinate_clause_ratio"] = (
        subordinate_count / sentence_count
        if sentence_count else 0
    )

    # --------------------------------------------------
    # Dependency structure
    # --------------------------------------------------

    dependency_counts = {}

    for token in doc:
        dependency_counts[token.dep_] = (
            dependency_counts.get(token.dep_, 0) + 1
        )

    features["root_count"] = dependency_counts.get(
        "ROOT", 0
    )

    features["subject_count"] = (
        dependency_counts.get("nsubj", 0)
        + dependency_counts.get("nsubjpass", 0)
    )

    features["object_count"] = (
        dependency_counts.get("dobj", 0)
        + dependency_counts.get("obj", 0)
    )

    # --------------------------------------------------
    # Sentence completeness proxies
    # --------------------------------------------------

    complete_sentence_count = 0

    for sent in sentences:

        has_subject = any(
            token.dep_ in {"nsubj", "nsubjpass", "expl"}
            for token in sent
        )

        has_predicate = any(
            token.pos_ in {"VERB", "AUX"}
            for token in sent
        )

        if has_subject and has_predicate:
            complete_sentence_count += 1

    features["complete_sentence_ratio"] = (
        complete_sentence_count / sentence_count
        if sentence_count else 0
    )

    features["incomplete_sentence_ratio"] = (
        1 - features["complete_sentence_ratio"]
        if sentence_count else 0
    )

    # --------------------------------------------------
    # Repetition
    # --------------------------------------------------

    adjacent_repetition = sum(
        1 for i in range(1, len(words))
        if words[i] == words[i - 1]
    )

    features["adjacent_repetition"] = adjacent_repetition

    features["repetition_ratio"] = (
        adjacent_repetition / word_count
        if word_count else 0
    )

    # --------------------------------------------------
    # Fillers
    # --------------------------------------------------

    filler_words = {
        "um", "uh", "erm", "hmm",
        "like", "actually", "basically"
    }

    filler_count = sum(
        words.count(word)
        for word in filler_words
    )

    features["filler_count"] = filler_count

    features["filler_ratio"] = (
        filler_count / word_count
        if word_count else 0
    )

    # --------------------------------------------------
    # Self-correction
    # --------------------------------------------------

    text_lower = text.lower()

    correction_phrases = [
        "i mean",
        "or rather",
        "sorry",
        "what i mean",
        "let me rephrase"
    ]

    correction_count = sum(
        text_lower.count(phrase)
        for phrase in correction_phrases
    )

    features["self_correction_count"] = correction_count

    # --------------------------------------------------
    # Function-word usage
    # --------------------------------------------------

    function_words = {
        "the", "a", "an",
        "is", "am", "are",
        "was", "were",
        "to", "of", "in",
        "on", "for", "with"
    }

    function_word_count = sum(
        word in function_words
        for word in words
    )

    features["function_word_ratio"] = (
        function_word_count / word_count
        if word_count else 0
    )

    # --------------------------------------------------
    # Sentence length categories
    # --------------------------------------------------

    if sentence_lengths:

        features["short_sentence_ratio"] = (
            np.mean(np.array(sentence_lengths) < 5)
        )

        features["medium_sentence_ratio"] = (
            np.mean(
                (np.array(sentence_lengths) >= 5) &
                (np.array(sentence_lengths) <= 20)
            )
        )

        features["long_sentence_ratio"] = (
            np.mean(np.array(sentence_lengths) > 20)
        )

    else:

        features["short_sentence_ratio"] = 0
        features["medium_sentence_ratio"] = 0
        features["long_sentence_ratio"] = 0

    return features

In [31]:
features = extract_advanced_linguistic_features(
    train_transcripts_df.iloc[0]["transcript"]
)

for key, value in features.items():
    print(f"{key}: {value}")

word_count: 91
unique_word_count: 61
type_token_ratio: 0.6703296703296703
sentence_count: 9
avg_sentence_length: 10.11111111111111
sentence_length_std: 4.6534203536382055
max_sentence_length: 22
min_sentence_length: 5
noun_count: 21
noun_ratio: 0.23076923076923078
verb_count: 9
verb_ratio: 0.0989010989010989
aux_count: 3
aux_ratio: 0.03296703296703297
adj_count: 9
adj_ratio: 0.0989010989010989
adv_count: 5
adv_ratio: 0.054945054945054944
pron_count: 10
pron_ratio: 0.10989010989010989
det_count: 12
det_ratio: 0.13186813186813187
adp_count: 11
adp_ratio: 0.12087912087912088
cconj_count: 4
cconj_ratio: 0.04395604395604396
sconj_count: 3
sconj_ratio: 0.03296703296703297
num_count: 0
num_ratio: 0.0
finite_verb_count: 14
verbs_per_sentence: 1.0
subordinate_clause_count: 6
subordinate_clause_ratio: 0.6666666666666666
root_count: 9
subject_count: 10
object_count: 3
complete_sentence_ratio: 0.8888888888888888
incomplete_sentence_ratio: 0.11111111111111116
adjacent_repetition: 1
repetition_ratio

In [35]:
import librosa
import numpy as np
import pandas as pd


def extract_acoustic_features(audio_path, sr=16000):

    y, _ = librosa.load(
        audio_path,
        sr=sr,
        mono=True
    )

    features = {}

    # --------------------------------------------------
    # Duration
    # --------------------------------------------------

    duration = len(y) / sr

    features["audio_duration"] = duration

    # --------------------------------------------------
    # RMS energy
    # --------------------------------------------------

    rms = librosa.feature.rms(y=y)[0]

    features["rms_mean"] = np.mean(rms)
    features["rms_std"] = np.std(rms)
    features["rms_min"] = np.min(rms)
    features["rms_max"] = np.max(rms)

    # --------------------------------------------------
    # Zero crossing rate
    # --------------------------------------------------

    zcr = librosa.feature.zero_crossing_rate(y)[0]

    features["zcr_mean"] = np.mean(zcr)
    features["zcr_std"] = np.std(zcr)

    # --------------------------------------------------
    # Spectral features
    # --------------------------------------------------

    spectral_centroid = librosa.feature.spectral_centroid(
        y=y,
        sr=sr
    )[0]

    spectral_bandwidth = librosa.feature.spectral_bandwidth(
        y=y,
        sr=sr
    )[0]

    spectral_rolloff = librosa.feature.spectral_rolloff(
        y=y,
        sr=sr
    )[0]

    features["spectral_centroid_mean"] = np.mean(
        spectral_centroid
    )

    features["spectral_centroid_std"] = np.std(
        spectral_centroid
    )

    features["spectral_bandwidth_mean"] = np.mean(
        spectral_bandwidth
    )

    features["spectral_bandwidth_std"] = np.std(
        spectral_bandwidth
    )

    features["spectral_rolloff_mean"] = np.mean(
        spectral_rolloff
    )

    features["spectral_rolloff_std"] = np.std(
        spectral_rolloff
    )

    # --------------------------------------------------
    # MFCC
    # --------------------------------------------------

    mfcc = librosa.feature.mfcc(
        y=y,
        sr=sr,
        n_mfcc=13
    )

    for i in range(13):

        features[f"mfcc_{i+1}_mean"] = np.mean(
            mfcc[i]
        )

        features[f"mfcc_{i+1}_std"] = np.std(
            mfcc[i]
        )

    # --------------------------------------------------
    # MFCC delta
    # --------------------------------------------------

    mfcc_delta = librosa.feature.delta(mfcc)

    mfcc_delta2 = librosa.feature.delta(
        mfcc,
        order=2
    )

    for i in range(13):

        features[f"mfcc_delta_{i+1}_mean"] = np.mean(
            mfcc_delta[i]
        )

        features[f"mfcc_delta_{i+1}_std"] = np.std(
            mfcc_delta[i]
        )

        features[f"mfcc_delta2_{i+1}_mean"] = np.mean(
            mfcc_delta2[i]
        )

        features[f"mfcc_delta2_{i+1}_std"] = np.std(
            mfcc_delta2[i]
        )

    # --------------------------------------------------
    # Speech / silence segmentation
    # --------------------------------------------------

    intervals = librosa.effects.split(
        y,
        top_db=35
    )

    if len(intervals) > 0:

        speech_samples = np.sum(
            intervals[:, 1] - intervals[:, 0]
        )

        speech_duration = speech_samples / sr

        silence_duration = max(
            duration - speech_duration,
            0
        )

        features["speech_duration"] = speech_duration
        features["silence_duration"] = silence_duration

        features["speech_ratio"] = (
            speech_duration / duration
            if duration > 0 else 0
        )

        features["silence_ratio"] = (
            silence_duration / duration
            if duration > 0 else 0
        )

        features["speech_segments"] = len(intervals)

        # --------------------------------------------------
        # Pause statistics
        # --------------------------------------------------

        if len(intervals) > 1:

            pauses = (
                intervals[1:, 0]
                - intervals[:-1, 1]
            ) / sr

            features["pause_count"] = len(pauses)

            features["pause_mean"] = np.mean(pauses)

            features["pause_std"] = np.std(pauses)

            features["pause_min"] = np.min(pauses)

            features["pause_max"] = np.max(pauses)

        else:

            features["pause_count"] = 0
            features["pause_mean"] = 0
            features["pause_std"] = 0
            features["pause_min"] = 0
            features["pause_max"] = 0

    else:

        features["speech_duration"] = 0
        features["silence_duration"] = duration
        features["speech_ratio"] = 0
        features["silence_ratio"] = 1
        features["speech_segments"] = 0
        features["pause_count"] = 0
        features["pause_mean"] = 0
        features["pause_std"] = 0
        features["pause_min"] = 0
        features["pause_max"] = 0

    # --------------------------------------------------
    # Approximate speaking rate
    # --------------------------------------------------

    features["speech_density"] = (
        duration
    )

    return features

In [33]:
import sys
!{sys.executable} -m pip install -q librosa


In [34]:
import librosa
import numpy as np
import pandas as pd

print("librosa version:", librosa.__version__)
print("NumPy version:", np.__version__)

librosa version: 1.0.0
NumPy version: 2.4.4


In [36]:
test_acoustic_features = []

for filename in tqdm(
    test_df["filename"],
    desc="Extracting test acoustic features"
):

    audio_path = os.path.join(
        TEST_AUDIO_DIR,
        filename
    )

    features = extract_acoustic_features(
        audio_path
    )

    test_acoustic_features.append(features)

test_acoustic_df = pd.DataFrame(
    test_acoustic_features
)

print(
    "Test acoustic feature shape:",
    test_acoustic_df.shape
)

Extracting test acoustic features:   0%|          | 0/216 [00:00<?, ?it/s]

Test acoustic feature shape: (216, 102)


In [37]:
import pandas as pd

# ============================================================
# 1. Add filename to feature DataFrames
# ============================================================

train_linguistic_df = train_linguistic_df.copy()
test_linguistic_df = test_linguistic_df.copy()

train_acoustic_df = train_acoustic_df.copy()
test_acoustic_df = test_acoustic_df.copy()

train_linguistic_df["filename"] = train_df["filename"].values
test_linguistic_df["filename"] = test_df["filename"].values

train_acoustic_df["filename"] = train_df["filename"].values
test_acoustic_df["filename"] = test_df["filename"].values


# ============================================================
# 2. Merge linguistic + acoustic features
# ============================================================

train_features_df = pd.merge(
    train_linguistic_df,
    train_acoustic_df,
    on="filename",
    how="inner"
)

test_features_df = pd.merge(
    test_linguistic_df,
    test_acoustic_df,
    on="filename",
    how="inner"
)


# ============================================================
# 3. Merge training labels
# ============================================================

train_features_df = pd.merge(
    train_features_df,
    train_df[["filename", "label"]],
    on="filename",
    how="inner"
)


# ============================================================
# 4. Check the final datasets
# ============================================================

print("Train shape:", train_features_df.shape)
print("Test shape :", test_features_df.shape)

print("\nTrain columns:", len(train_features_df.columns))
print("Test columns :", len(test_features_df.columns))

print("\nFirst 5 training samples:")
display(train_features_df.head())

print("\nFirst 5 test samples:")
display(test_features_df.head())

NameError: name 'train_linguistic_df' is not defined

In [38]:
import os
import pandas as pd
import numpy as np

TRANSCRIPT_DIR = "/Users/komal/Documents/SHL/transcripts"

TRAIN_TRANSCRIPT_CSV = os.path.join(
    TRANSCRIPT_DIR, "train_transcripts.csv"
)

TEST_TRANSCRIPT_CSV = os.path.join(
    TRANSCRIPT_DIR, "test_transcripts.csv"
)

train_transcripts_df = pd.read_csv(TRAIN_TRANSCRIPT_CSV)
test_transcripts_df = pd.read_csv(TEST_TRANSCRIPT_CSV)

print("Train transcripts:", train_transcripts_df.shape)
print("Test transcripts :", test_transcripts_df.shape)

display(train_transcripts_df.head())
display(test_transcripts_df.head())

Train transcripts: (769, 3)
Test transcripts : (216, 2)


,filename,label,transcript
0,audio_192.wav,3.0,"Well, this petite land is very, very beautiful..."
1,audio_436.wav,3.0,"To see no of a school playground, our playgrou..."
2,audio_447.wav,3.5,I favorite place to visit is the beach. I love...
3,audio_126.wav,2.0,"Once upon a time, we, with five friends, went ..."
4,audio_61.wav,2.0,"Well, when I was a child, I remember I went to..."


,filename,transcript
0,audio_128.wav,The moment in my life that I cherish the most ...
1,audio_156.wav,One of the role models is someone who embadiys...
2,audio_19.wav,"I guess my favorite sport would be Taekwondo, ..."
3,audio_108.wav,Hospital having more of doctors and there are ...
4,audio_206.wav,"I remember when I went to Goa for a trip, then..."


In [39]:
import spacy

nlp = spacy.load("en_core_web_sm")

print("spaCy model loaded successfully")

spaCy model loaded successfully


In [40]:
from tqdm.auto import tqdm

tqdm.pandas()

# ============================================================
# TRAIN linguistic features
# ============================================================

train_linguistic_features = []

for text in tqdm(
    train_transcripts_df["transcript"].fillna(""),
    desc="Train linguistic features"
):
    train_linguistic_features.append(
        extract_advanced_linguistic_features(text)
    )

train_linguistic_df = pd.DataFrame(
    train_linguistic_features
)

train_linguistic_df["filename"] = (
    train_transcripts_df["filename"].values
)


# ============================================================
# TEST linguistic features
# ============================================================

test_linguistic_features = []

for text in tqdm(
    test_transcripts_df["transcript"].fillna(""),
    desc="Test linguistic features"
):
    test_linguistic_features.append(
        extract_advanced_linguistic_features(text)
    )

test_linguistic_df = pd.DataFrame(
    test_linguistic_features
)

test_linguistic_df["filename"] = (
    test_transcripts_df["filename"].values
)


print("Train linguistic shape:",
      train_linguistic_df.shape)

print("Test linguistic shape:",
      test_linguistic_df.shape)

Train linguistic features:   0%|          | 0/769 [00:00<?, ?it/s]

Test linguistic features:   0%|          | 0/216 [00:00<?, ?it/s]

Train linguistic shape: (769, 49)
Test linguistic shape: (216, 49)


In [41]:
print("train_acoustic_df exists:",
      "train_acoustic_df" in globals())

print("test_acoustic_df exists:",
      "test_acoustic_df" in globals())

train_acoustic_df exists: False
test_acoustic_df exists: True


In [42]:
from tqdm.auto import tqdm
import os
import pandas as pd

# ============================================================
# Extract TRAIN acoustic features
# ============================================================

train_acoustic_features = []

for filename in tqdm(
    train_transcripts_df["filename"],
    desc="Extracting train acoustic features"
):
    audio_path = os.path.join(
        TRAIN_AUDIO_DIR,
        filename
    )

    features = extract_acoustic_features(audio_path)

    train_acoustic_features.append(features)


# Convert to DataFrame
train_acoustic_df = pd.DataFrame(
    train_acoustic_features
)

# Add filename for safe merging
train_acoustic_df["filename"] = (
    train_transcripts_df["filename"].values
)


print("Train acoustic feature shape:",
      train_acoustic_df.shape)

display(train_acoustic_df.head())

Extracting train acoustic features:   0%|          | 0/769 [00:00<?, ?it/s]

Train acoustic feature shape: (769, 103)


,audio_duration,rms_mean,rms_std,rms_min,rms_max,zcr_mean,zcr_std,spectral_centroid_mean,spectral_centroid_std,spectral_bandwidth_mean,...,speech_ratio,silence_ratio,speech_segments,pause_count,pause_mean,pause_std,pause_min,pause_max,speech_density,filename
0,60.508375,0.031786,0.038404,0.000104,0.188842,0.084768,0.053518,979.478539,399.567888,916.472067,...,0.600247,0.399753,36,35,0.569600,0.658067,0.032,2.688,60.508375,audio_192.wav
1,60.074688,0.068651,0.055423,0.000282,0.400792,0.142977,0.127519,1813.837012,1156.460316,1502.301143,...,0.897727,0.102273,40,39,0.123897,0.214746,0.032,1.376,60.074688,audio_436.wav
2,60.074688,0.054421,0.051533,0.000957,0.249218,0.146870,0.129777,1652.475254,1207.838332,1500.652559,...,0.803267,0.196733,41,40,0.276000,0.398377,0.032,2.432,60.074688,audio_447.wav
3,45.060000,0.013778,0.012183,0.000037,0.061771,0.059477,0.036870,996.631363,471.566967,1177.252185,...,0.864359,0.135641,30,29,0.210759,0.127313,0.032,0.448,45.060000,audio_126.wav
4,45.300000,0.009330,0.008970,0.000003,0.042700,0.169988,0.128430,2278.144095,1068.412091,1630.415181,...,0.730861,0.269139,27,26,0.352000,0.224000,0.032,0.832,45.300000,audio_61.wav


In [44]:
# ============================================================
# Add filenames to acoustic feature DataFrames
# ============================================================

train_acoustic_df["filename"] = (
    train_transcripts_df["filename"].values
)

test_acoustic_df["filename"] = (
    test_transcripts_df["filename"].values
)

print("Train acoustic shape:", train_acoustic_df.shape)
print("Test acoustic shape :", test_acoustic_df.shape)

print("\nTrain filenames unique:",
      train_acoustic_df["filename"].is_unique)

print("Test filenames unique:",
      test_acoustic_df["filename"].is_unique)

Train acoustic shape: (769, 103)
Test acoustic shape : (216, 103)

Train filenames unique: True
Test filenames unique: True


In [45]:
print("Train linguistic:", train_linguistic_df.shape)
print("Test linguistic :", test_linguistic_df.shape)

print("\nTrain linguistic filename:",
      "filename" in train_linguistic_df.columns)

print("Test linguistic filename:",
      "filename" in test_linguistic_df.columns)

Train linguistic: (769, 49)
Test linguistic : (216, 49)

Train linguistic filename: True
Test linguistic filename: True


In [46]:
# Make sure test has filename
if "filename" not in test_acoustic_df.columns:
    test_acoustic_df["filename"] = (
        test_transcripts_df["filename"].values
    )

# Merge linguistic + acoustic
train_features_df = pd.merge(
    train_linguistic_df,
    train_acoustic_df,
    on="filename",
    how="inner"
)

test_features_df = pd.merge(
    test_linguistic_df,
    test_acoustic_df,
    on="filename",
    how="inner"
)

# Add training labels
train_features_df = pd.merge(
    train_features_df,
    train_transcripts_df[["filename", "label"]],
    on="filename",
    how="inner"
)

print("Train:", train_features_df.shape)
print("Test :", test_features_df.shape)

Train: (769, 152)
Test : (216, 151)


In [47]:
# Check that all training samples are present
assert len(train_features_df) == len(train_transcripts_df), \
    "Training samples were lost during merging!"

assert len(test_features_df) == len(test_transcripts_df), \
    "Test samples were lost during merging!"

# Check labels
label_check = train_features_df[
    ["filename", "label"]
].merge(
    train_transcripts_df[["filename", "label"]],
    on="filename",
    suffixes=("_merged", "_original")
)

assert (
    label_check["label_merged"]
    == label_check["label_original"]
).all(), "Label mismatch detected!"

print("✓ All 769 training samples preserved")
print("✓ All 216 test samples preserved")
print("✓ Labels correctly aligned")

✓ All 769 training samples preserved
✓ All 216 test samples preserved
✓ Labels correctly aligned


In [81]:
import sys

!{sys.executable} -m pip install -q sentence-transformers

In [82]:
from sentence_transformers import SentenceTransformer

embedding_model = SentenceTransformer(
    "all-MiniLM-L6-v2"
)

print("Embedding model loaded.")

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

Embedding model loaded.


In [83]:
train_transcripts_df = pd.read_csv(
    "/Users/komal/Documents/SHL/transcripts/train_transcripts.csv"
)

test_transcripts_df = pd.read_csv(
    "/Users/komal/Documents/SHL/transcripts/test_transcripts.csv"
)

train_text = (
    train_transcripts_df["transcript"]
    .fillna("")
    .astype(str)
    .tolist()
)

test_text = (
    test_transcripts_df["transcript"]
    .fillna("")
    .astype(str)
    .tolist()
)

In [84]:
train_embeddings = embedding_model.encode(
    train_text,
    batch_size=32,
    show_progress_bar=True,
    normalize_embeddings=True
)

test_embeddings = embedding_model.encode(
    test_text,
    batch_size=32,
    show_progress_bar=True,
    normalize_embeddings=True
)

print("Train embeddings:", train_embeddings.shape)
print("Test embeddings :", test_embeddings.shape)

Batches:   0%|          | 0/25 [00:00<?, ?it/s]

Batches:   0%|          | 0/7 [00:00<?, ?it/s]

Train embeddings: (769, 384)
Test embeddings : (216, 384)


In [85]:
X
y
X_test

,word_count,unique_word_count,type_token_ratio,sentence_count,avg_sentence_length,sentence_length_std,max_sentence_length,min_sentence_length,noun_count,noun_ratio,...,silence_ratio,speech_segments,pause_count,pause_mean,pause_std,pause_min,pause_max,speech_density,words_per_second,words_per_minute
0,93,48,0.516129,5,18.600000,3.826225,22,13,16,0.172043,...,0.282024,30,29,0.375172,0.433246,0.032,1.632,45.060000,2.063915,123.834887
1,68,57,0.838235,4,17.000000,7.968689,29,8,22,0.323529,...,0.303347,27,26,0.491077,0.484162,0.032,1.856,44.969313,1.512142,90.728538
2,104,66,0.634615,7,14.857143,6.311635,25,6,16,0.153846,...,0.190971,33,32,0.265000,0.334782,0.032,1.600,45.012000,2.310495,138.629699
3,54,42,0.777778,4,13.500000,6.837397,20,2,12,0.222222,...,0.183927,53,52,0.193846,0.215067,0.032,0.896,58.980000,0.915565,54.933876
4,85,51,0.600000,4,21.250000,19.954636,55,6,16,0.188235,...,0.002131,1,0,0.000000,0.000000,0.000,0.000,45.040000,1.887211,113.232682
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
211,108,57,0.527778,6,18.000000,5.537749,26,9,22,0.203704,...,0.031247,11,10,0.140800,0.078645,0.032,0.288,45.060000,2.396804,143.808256
212,76,33,0.434211,9,8.444444,3.774508,15,5,12,0.157895,...,0.000000,1,0,0.000000,0.000000,0.000,0.000,45.060000,1.686640,101.198402
213,90,59,0.655556,8,11.250000,4.993746,19,4,10,0.111111,...,0.105903,23,22,0.209455,0.147506,0.032,0.544,45.060000,1.997337,119.840213
214,73,48,0.657534,7,10.428571,8.243736,28,1,15,0.205479,...,0.150704,22,21,0.260571,0.246151,0.032,0.992,41.220000,1.770985,106.259098


In [86]:
embedding_train_df = pd.DataFrame(
    train_embeddings,
    columns=[
        f"emb_{i}"
        for i in range(train_embeddings.shape[1])
    ]
)

embedding_train_df["filename"] = (
    train_transcripts_df["filename"].values
)

embedding_test_df = pd.DataFrame(
    test_embeddings,
    columns=[
        f"emb_{i}"
        for i in range(test_embeddings.shape[1])
    ]
)

embedding_test_df["filename"] = (
    test_transcripts_df["filename"].values
)

In [87]:
multimodal_train = pd.merge(
    train_features_df,
    embedding_train_df,
    on="filename",
    how="inner"
)

multimodal_test = pd.merge(
    test_features_df,
    embedding_test_df,
    on="filename",
    how="inner"
)

print("Train:", multimodal_train.shape)
print("Test :", multimodal_test.shape)

Train: (769, 538)
Test : (216, 537)


In [88]:
X_multi = multimodal_train.drop(
    columns=["filename", "label"]
).copy()

y_multi = multimodal_train["label"].copy()

X_multi_test = multimodal_test.drop(
    columns=["filename"]
).copy()

print("X:", X_multi.shape)
print("y:", y_multi.shape)
print("Test:", X_multi_test.shape)

X: (769, 536)
y: (769,)
Test: (216, 536)


In [98]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import (
    Input,
    Dense,
    Dropout,
    BatchNormalization
)


def build_multimodal_mlp(input_dim):

    model = Sequential([

        Input(shape=(input_dim,)),

        Dense(256, activation="relu"),
        BatchNormalization(),
        Dropout(0.30),

        Dense(128, activation="relu"),
        BatchNormalization(),
        Dropout(0.25),

        Dense(64, activation="relu"),
        Dropout(0.20),

        Dense(32, activation="relu"),

        Dense(1, activation="linear")
    ])

    model.compile(
        optimizer=tf.keras.optimizers.Adam(
            learning_rate=0.0005
        ),
        loss="mse"
    )

    return model

In [99]:
from sklearn.model_selection import KFold
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
from scipy.stats import pearsonr

kf = KFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

multi_oof = np.zeros(len(y_multi))

fold_rmse = []
fold_pearson = []

for fold, (train_idx, val_idx) in enumerate(
    kf.split(X_multi),
    start=1
):

    print(f"\n========== FOLD {fold} ==========")

    X_train = X_multi.iloc[train_idx]
    X_val = X_multi.iloc[val_idx]

    y_train = y_multi.iloc[train_idx].values
    y_val = y_multi.iloc[val_idx].values

    # ---------------------------------------
    # Scaling
    # ---------------------------------------

    scaler = StandardScaler()

    X_train_scaled = scaler.fit_transform(X_train)
    X_val_scaled = scaler.transform(X_val)

    # ---------------------------------------
    # Model
    # ---------------------------------------

    model = build_multimodal_mlp(
        X_train_scaled.shape[1]
    )

    early_stopping = EarlyStopping(
        monitor="val_loss",
        patience=25,
        restore_best_weights=True
    )

    reduce_lr = ReduceLROnPlateau(
        monitor="val_loss",
        factor=0.5,
        patience=8,
        min_lr=1e-6
    )

    # ---------------------------------------
    # Training
    # ---------------------------------------

    history = model.fit(
        X_train_scaled,
        y_train,
        validation_data=(
            X_val_scaled,
            y_val
        ),
        epochs=250,
        batch_size=32,
        callbacks=[
            early_stopping,
            reduce_lr
        ],
        verbose=0
    )

    # ---------------------------------------
    # Prediction
    # ---------------------------------------

    pred = model.predict(
        X_val_scaled,
        verbose=0
    ).ravel()

    pred = np.clip(
        pred,
        0,
        5
    )

    multi_oof[val_idx] = pred

    # ---------------------------------------
    # Metrics
    # ---------------------------------------

    rmse = np.sqrt(
        mean_squared_error(
            y_val,
            pred
        )
    )

    pearson = pearsonr(
        y_val,
        pred
    )[0]

    fold_rmse.append(rmse)
    fold_pearson.append(pearson)

    print(f"RMSE    : {rmse:.4f}")
    print(f"Pearson : {pearson:.4f}")
    print(f"Epochs  : {len(history.history['loss'])}")


========== FOLD 1 ==========
RMSE    : 0.9026
Pearson : 0.8055
Epochs  : 94

========== FOLD 2 ==========
RMSE    : 0.8014
Pearson : 0.7925
Epochs  : 57

========== FOLD 3 ==========
RMSE    : 1.0444
Pearson : 0.7455
Epochs  : 56

========== FOLD 4 ==========
RMSE    : 0.8365
Pearson : 0.8050
Epochs  : 58

========== FOLD 5 ==========
RMSE    : 0.8600
Pearson : 0.7509
Epochs  : 77


In [100]:
overall_rmse = np.sqrt(
    mean_squared_error(
        y_multi,
        multi_oof
    )
)

overall_pearson = pearsonr(
    y_multi,
    multi_oof
)[0]

print("\n======================================")
print("MULTIMODAL MODEL — OOF RESULTS")
print("======================================")

print(
    f"Overall OOF RMSE    : {overall_rmse:.4f}"
)

print(
    f"Overall OOF Pearson : {overall_pearson:.4f}"
)

print(
    f"Mean Fold RMSE      : {np.mean(fold_rmse):.4f}"
)

print(
    f"Mean Fold Pearson   : {np.mean(fold_pearson):.4f}"
)


MULTIMODAL MODEL — OOF RESULTS
Overall OOF RMSE    : 0.8930
Overall OOF Pearson : 0.7770
Mean Fold RMSE      : 0.8890
Mean Fold Pearson   : 0.7799


# 1. Approach
The objective of this project is to build an end-to-end multimodal regression pipeline to predict spoken language proficiency and grammar scores from audio recordings. To achieve high predictive accuracy, the model combines three complementary data modalities:

Linguistic Features: Captures vocabulary richness, syntactic complexity, Part-of-Speech (POS) distributions, and fluency markers (fillers and repetitions).

Acoustic Features: Measures prosody, voice energy, spectral properties, and speech-to-silence segmentation (pauses and pacing).

Deep Semantic Embeddings: Utilizes transformer-based sentence embeddings to capture deep contextual semantic meaning from the transcripts.

# 2. Preprocessing Steps
Dataset Extraction: Unzipped and structured the dataset containing training metadata (train.csv, 769 samples), test metadata (test.csv, 216 samples), and respective audio directories.

Audio Standardization: Loaded all audio files using librosa at a standardized sampling rate of 16 kHz in mono.

Speech-to-Text Transcription: Converted raw audio waveforms into written text transcripts using OpenAI's Whisper (base model).

NLP & Linguistic Feature Extraction: Used spaCy (en_core_web_sm) for tokenization, POS tagging, and syntactic analysis to engineer 49 linguistic variables.

Semantic Vectorization: Generated normalized dense vector representations (384 dimensions) of the transcripts using the Sentence Transformer model (all-MiniLM-L6-v2).

Acoustic Feature Extraction: Extracted ~102 acoustic descriptors via librosa, including RMS energy, spectral centroids/bandwidth, 13 Mel-Frequency Cepstral Coefficients (MFCCs) with deltas, and pause/silence statistics.

# 3. Pipeline Architecture
Feature Integration: Merged linguistic features, acoustic features, and Sentence Transformer embeddings into a unified tabular matrix using the filename key.

Validation Strategy: Implemented 5-Fold Cross-Validation (KFold with shuffling and random_state=42) to robustly train and validate the model.

Data Scaling: Applied StandardScaler independently across training and validation folds to normalize input features.

Neural Network Architecture (Multilayer Perceptron):

Input Layer: Dimension matching the combined multimodal feature space.

Hidden Layers: Dense layer (256 units, ReLU activation), Batch Normalization, Dropout (0.3), Dense layer (128 units, ReLU) , Batch Normalization, Dropout (0.3)

Output Layer: Single dense unit (linear activation) with predictions clipped between 0 and 5.

Optimization & Callbacks: Trained using Adam optimizer with EarlyStopping (patience=25) and ReduceLROnPlateau (factor=0.5, patience=8).

# 4. Evaluation Results

The model was evaluated using Root Mean Squared Error (RMSE) and Pearson Correlation Coefficient across all 5 folds:

Fold 1: RMSE = 0.9026, Pearson = 0.8055 (94 epochs)

Fold 2: RMSE = 0.8014, Pearson = 0.7925 (57 epochs)

Fold 3: RMSE = 1.0444, Pearson = 0.7455 (56 epochs)

Fold 4: RMSE = 0.8365, Pearson = 0.8050 (58 epochs)

Fold 5: RMSE = 0.8600, Pearson = 0.7509 (77 epochs)

Aggregated Performance Metrics

Overall Out-Of-Fold (OOF) RMSE: 0.8930

Overall OOF Pearson Correlation: 0.7770

Mean Fold RMSE: 0.8890

Mean Fold Pearson Correlation: 0.7799
